amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/PeriodicShardSyncManager.java

```diff
@@ -116,24 +154,244 @@ public void stop() {
+                            leftLeaseHashRange.endingHashKey().max(rightLeaseHashRange.endingHashKey()));
+                } else {
+                    // Case of non overlapping leases when rangediff is positive. signum() will be 1 for positive.
+                    // If rangeDiff is 1, then it is a case of continuous hashrange. If not, it is a hole.
+                    if (!rangeDiff.equals(BigInteger.ONE)) {
+                        log.error("Incomplete hash range found for {} between {} and {}.", streamIdentifier,
+                                leftMostLeaseToReportInCaseOfHole, sortedLeasesWithHashKeyRanges.get(i));
+                        return Optional.of(new HashRangeHole(leftMostLeaseToReportInCaseOfHole,
+                                sortedLeasesWithHashKeyRanges.get(i)));
+                    }
+                    leftMostLeaseToReportInCaseOfHole = sortedLeasesWithHashKeyRanges.get(i);
+                    leftLeaseHashRange = rightLeaseHashRange;
+                }
+            }
+        }
+        return Optional.empty();
+    }
+
+    @VisibleForTesting
+    static List<Lease> sortLeasesByHashRange(List<Lease> leasesWithHashKeyRanges) {
+        if (leasesWithHashKeyRanges.size() == 0 || leasesWithHashKeyRanges.size() == 1)
+            return leasesWithHashKeyRanges;
+        Collections.sort(leasesWithHashKeyRanges, new HashKeyRangeComparator());
+        return leasesWithHashKeyRanges;
+    }
+
+    @Value
+    private static class HashRangeHole {
+        HashRangeHole() {
+            leaseAtEndOfPossibleHole = leaseAtStartOfPossibleHole = null;
+        }
+
+        HashRangeHole(Lease leaseAtStartOfPossibleHole, Lease leaseAtEndOfPossibleHole) {
+            this.leaseAtStartOfPossibleHole = leaseAtStartOfPossibleHole;
+            this.leaseAtEndOfPossibleHole = leaseAtEndOfPossibleHole;
+        }
+
+        private final Lease leaseAtStartOfPossibleHole;
+        private final Lease leaseAtEndOfPossibleHole;
+    }
+
+    private static class HashRangeHoleTracker {
+        private HashRangeHole hashRangeHole;
+        @Getter
+        private Integer numConsecutiveHoles;
+
+        public boolean hasHighConfidenceOfHoleWith(@NonNull HashRangeHole hashRangeHole) {
+            if (hashRangeHole.equals(this.hashRangeHole)) {
+                ++this.numConsecutiveHoles;
+            } else {
+                this.hashRangeHole = hashRangeHole;
+                this.numConsecutiveHoles = 1;
+            }
+            return numConsecutiveHoles >= CONSECUTIVE_HOLES_FOR_TRIGGERING_RECOVERY;
         }
     }
 
     /**
-     * Checks if the entire hash range is covered
-     * @return true if covered, false otherwise
+     * Helper class to compare leases based on their hash range.
      */
-    public boolean hashRangeCovered() {
-        // TODO: Implement method
-        return true;
+    private static class HashKeyRangeComparator implements Comparator<Lease>, Serializable {
```

Comment:
I think the comparator should have comparison on the ending hash key range... if the start are same otherwise we might report inconsistent hole. May be we can add test cases for the below, if not done

Ex : 
Assume :  0 ---> 100 is the hash range possible and  following is the shard map of a customer.
.................. s1
................. /...\
...............s2...s3
..................... /..\
....................s4..s5
..................... \.../
.......................s6

shard1 => 0->100 (SHARD_END)
shard2 => 0->50(IN_PROGRESS)    ;  shard3=> 51 ->100 (SHARD_END)
                               ;  shard4=> 51->75 (IN_PROGRESS) ; shard5=>76->100(DONE);
                               ;  shard6=> 51->100 (BLOCKED_ON_PARENT)

Assume shard 5 is ended, if we do sorting on starting hash ranges for all leases which are in progress shard4 and shard 6 as same start hash key and we should always have shard6 as the last to validate the hash range completeness otherwise we might report inconsistencies.

Reply:
This is not an inconsistency right? We will see 0-50,51-100 as complete hashrange and we will report hashrangecomplete. While we see 51-75, that will be ignored as part of the merge.

Reply:
This is addressed after adding the comparison on the ending hash key otherwise we might report this shardmap as inconsistent